Define RSTRING_SUCCESS global constant

sukibaby opened the pull request:
It isn't immediately obvious that an empty RString is used throughout the codebase to indicate success, so this constant makes its purpose more clear. The purpose of the empty RString is documented in very few places, so unless you happen to find one of them, it's not at all clear what the empty RString is doing.

sukibaby committed:
Define RSTRING_SUCCESS global constant

It isn't immediately obvious that an empty RString is used throughout the codebase to indicate success, so this constant makes its purpose more clear.

Cases where an empty RString indicates failure are left alone but annotated.

Revert "Define RSTRING_SUCCESS global constant"

This reverts commit 473c1fcb6734fd41e736aee5d35d63a3f488ff9b.

a

teejusb commented on src/global.h:
```diff
@@ -107,7 +107,7 @@ void ShowWarningOrTrace( const char *file, int line, const char *message, bool b
 
 #include "RageException.h"
-
+const RString SUCCESS_INDICATOR = RString();
```
Preference to call it `RSTRING_SUCCESS`.

florczakraf commented on src/AnnouncerManager.cpp:
```diff
@@ -114,7 +114,7 @@ static const char *aliases[][2] = {
 {
 	if(sAnnouncerName == "")
-		return RString(); /* announcer disabled */
+		return SUCCESS_INDICATOR; /* announcer disabled */
```
this is not a success exit

florczakraf commented on src/AnnouncerManager.cpp:
```diff
@@ -142,7 +142,7 @@ RString AnnouncerManager::GetPathTo( RString sAnnouncerName, RString sFolderName
 #endif
 
-	return RString();
+	return SUCCESS_INDICATOR;
```
neither is this one

florczakraf commented on src/Character.cpp:
```diff
@@ -80,7 +80,7 @@ RString GetRandomFileInDir( RString sDir )
 	GetDirListing( sDir, asFiles, false, true );
 	if( asFiles.empty() )
-		return RString();
+		return SUCCESS_INDICATOR;
```
or this one

florczakraf commented on src/Character.cpp:
```diff
@@ -91,7 +91,7 @@ RString Character::GetModelPath() const
 		return s;
 	else
-		return RString();
+		return SUCCESS_INDICATOR;
```
not a success

florczakraf commented on src/Character.cpp:
```diff
@@ -106,7 +106,7 @@ RString Character::GetTakingABreakPath() const
 	GetDirListing( m_sCharDir+"break.bmp", as, false, true );
 	if( as.empty() )
-		return RString();
+		return SUCCESS_INDICATOR;
```
not a success

florczakraf commented on src/Character.cpp:
```diff
@@ -131,7 +131,7 @@ RString Character::GetSongSelectIconPath() const
 		GetDirListing( m_sCharDir+"icon.bmp", as, false, true );
 		if( as.empty() )
-			return RString();
+			return SUCCESS_INDICATOR;
```
not a success

florczakraf reviewed (changes requested):
I started browsing this PR, but none of the changes I managed to process so far were indicating success -- returning an empty path to a file or missing screen name is quite the opposite, so using the common new name only introduces confusion. I think that you've replaced too many places -- I feel like you meant only to replace it in functions that are expected to return an error message or no message in case of success.

sukibaby commented:
> I started browsing this PR, but none of the changes I managed to process so far were indicating success -- returning an empty path to a file or missing screen name is quite the opposite, so using the common new name only introduces confusion. I think that you've replaced too many places -- I feel like you meant only to replace it in functions that are expected to return an error message or no message in case of success.

Thanks for checking this over! I'll likely accept your suggestions.

sukibaby commented:
1. Renamed `SUCCESS_INDICATOR` to `RSTRING_SUCCESS`
2. Reverted changes where an empty RString indicated failure or no file found, and annotated them as such

florczakraf commented:
> 2. Reverted changes where an empty RString indicated failure or no file found, and annotated them as such

I'm not sure if this is deliberate or not, but you only updated some places that I indicated before I gave up writing comments. There are still dozens of other ones that don't make any sense with the new `RSTRING_SUCCESS`

sukibaby commented:
> > 2. Reverted changes where an empty RString indicated failure or no file found, and annotated them as such
> 
> I'm not sure if this is deliberate or not, but you only updated some places that I indicated before I gave up writing comments. There are still dozens of other ones that don't make any sense with the new `RSTRING_SUCCESS`

It was deliberate, I do want to use the constant anywhere it is remotely appropriate, because I feel it's easier for a causal reader of the code to get an idea of what the empty RString could be used for, but I do acknowledge that's very subjective.

florczakraf commented:
> It was deliberate, I do want to use the constant anywhere it is remotely appropriate, because I feel it's easier for a causal reader of the code to get an idea of what the empty RString could be used for, but I do acknowledge that's very subjective.

If it's deliberate, then I have to assume you're just trolling right now. Not only it's "subjective" -- it's objectively wrong in places like:
```
		LOG->Warn( "GetMD5: Failed to open file '%s'", fn.c_str() );
		return RSTRING_SUCCESS;
```

or

```
	if( sError.size() != 0 )
		return RSTRING_SUCCESS;
```

and many more.

To be extra clear: I'm perfectly fine with using common `RSTRING_SUCCESS` **in functions that return error messages, or an empty message when there is no error to report**. However, using this anywhere else, even with a comment, and especially in the context of evident failure is much worse for the future readers, because it makes them question the code every time they stumble upon it during debugging -- "hey is this SUCCESS actually a success, or is it a silent failure? I'm expecting a path here but it returned a SUCCESS, what's up with that?" This is supposed to decrease the cognitive complexity of the code, not increase it -- please think of all the people that are going to work on this project in the future.

sukibaby commented:
> If it's deliberate, then I have to assume you're just trolling right now. Not only it's "subjective" -- it's objectively wrong in places like:

No need for this comment.  I think anyone would start glazing over the code after looking at that many occurences of RString().  

Some of them I had a hard time deciding because I feel it may be more ambiguous: let's say a loop to indicate a failure executes correctly. One would say that loop succeeded in doing its job. So do we call it a success or a failure? I have a hard time deciding if `RSTRING_SUCCESS` is better in that case. I wouldn't want to lead someone to believe that the function which successfully indicated a failure occurred is returning a success. Maybe that is getting too philosophical about the code but it's something I was trying to weigh in my mind.

There may be occasions like this where I felt `RSTRING_SUCCESS` was the right choice even when a failure is being returned. I readily accept that these may be too ambiguous to be worth applying `RSTRING_SUCCESS` to.

sukibaby commented:
Setting this to draft while I go thru it again - reviews are more than welcome in the meantime.

sukibaby commented:
Ok, this is all set now. I did another run thru and took out all the ambiguous or confusing ones.

sukibaby commented:
## Closing in favor of removing global.h
